document/core/exec/instructions.rst

```diff
@@ -254,23 +254,474 @@ Reference Instructions
+      \land & (\bytes_{\X{ft}}(i) = \unpackval_{\X{ft}}(b^\ast))^n)
+     \end{array} \\
+   \end{array}
+
+
+.. _exec-array.new_elem:
+
+:math:`\ARRAYNEWELEM~\typeidx~\elemidx`
+.......................................
+
+.. todo:: Prose
+
+.. math::
+   ~\\[-1ex]
+   \begin{array}{lcl@{\qquad}l}
+   S; F; (\I32.\CONST~s)~(\I32.\CONST~n)~(\ARRAYNEWELEM~x~y) &\stepto& \TRAP
+     \\&&
+     (\iff s + n > |S.\SELEMS[F.\AMODULE.\MIELEMS[y]].\EIELEM|)
+   \\[1ex]
+   S; F; (\I32.\CONST~s)~(\I32.\CONST~n)~(\ARRAYNEWELEM~x~y) &\stepto& \reff^n~(\ARRAYNEWFIXED~x)
+     \\&&
+     \begin{array}[t]{@{}r@{~}l@{}}
+      (\iff & \reff^n = S.\SELEMS[F.\AMODULE.\MIELEMS[y]].\EIELEM[s \slice n])
+     \end{array} \\
+   \end{array}
+
+
+.. _exec-array.get:
+.. _exec-array.get_sx:
+
+:math:`\ARRAYGET\K{\_}\sx^?~\typeidx`
+.....................................
+
+.. todo:: Prose
+
+.. math::
+   \begin{array}{lcl@{\qquad}l}
+   S; (\REFARRAYADDR~a)~(\I32.\CONST~i)~(\ARRAYGET\K{\_}\sx^?~x) &\stepto& \val
+     &
+     \begin{array}[t]{@{}r@{~}l@{}}
+      (\iff & F.\AMODULE.\MITYPES[x] = \TARRAY~\X{ft} \\
+      \land & \val = \unpackval^{\sx^?}_{\X{ft}}(S.\SARRAYS[a].\AIFIELDS[i]))
+     \end{array} \\
+   \end{array}
+
+
+.. _exec-array.set:
+
+:math:`\ARRAYSET~\typeidx`
+..........................
+
+.. todo:: Prose
+
+.. math::
+   \begin{array}{lcl@{\qquad}l}
+   S; (\REFARRAYADDR~a)~(\I32.\CONST~i)~\val~(\ARRAYSET~x) &\stepto& S'; \epsilon
+     &
+     \begin{array}[t]{@{}r@{~}l@{}}
+     (\iff & F.\AMODULE.\MITYPES[x] = \TSTRUCT~\X{ft}^n \\
+      \land & S' = S \with \SARRAYS[a].\AIFIELDS[i] = \packval_{\X{ft}}(\val))
+     \end{array} \\
+   \end{array}
+
+
+.. _exec-array.len:
+
+:math:`\ARRAYLEN`
+.................
+
+.. todo:: Prose
+
+.. math::
+   \begin{array}{lcl@{\qquad}l}
+   S; (\REFARRAYADDR~a)~\ARRAYLEN &\stepto& (\I32.\CONST~|\SARRAYS[a].\AIFIELDS[i]|)
+   \end{array}
+
+
+.. _exec-extern.externalize:
+
+:math:`\EXTERNEXTERNALIZE`
+..........................
+
+1. Assert: due to :ref:`validation <valid-extern.externalize>`, a :ref:`reference value <syntax-ref>` is on the top of the stack.
+
+2. Pop the value :math:`\reff` from the stack.
+
+3. Let :math:`\reff'` be the reference value :math:`(\REFEXTERN~\reff)`.
+
+5. Push the reference value :math:`\reff'` to the stack.
+
+.. math::
+   \begin{array}{lcl@{\qquad}l}
+   \reff~\EXTERNEXTERNALIZE &\stepto& (\REFEXTERN~\reff)
+   \end{array}
+
+
+.. _exec-extern.internalize:
+
+:math:`\EXTERNINTERNALIZE`
+..........................
+
+1. Assert: due to :ref:`validation <valid-extern.internalize>`, a :ref:`reference value <syntax-ref>` is on the top of the stack.
+
+2. Pop the value :math:`\reff` from the stack.
+
+3. Assert: due to :ref:`validation <valid-extern.internalize>`, a :math:`\reff` is an :ref:`external reference <syntax-ref.extern>`.
+
+4. Let :math:`(\REFEXTERN~\reff')` be the reference value :math:`\reff`.
```

Comment:
This is pattern matching saying reff = ref.extern ref'. Are you asking why we need the ref.extern constructor at all? The reason is that the rules for import type checking need to be able to deduce the type if each value, in such a way that we can distinguish internal from external representation (because mixing them must be rejected).

Reply:
It makes sense that we need the constructor, but it seems weird to use the same name on both sides of the equality, if that's what's happening. I'm reading this as `reff = ref.extern reff`, whereas I would expect to have something like `reff = ref.extern reff'` instead.